File: Data Structures & Algorithms/subsets/submission-0.java
class Solution {
    public static List<List<Integer>> dfs(int[] arr, int index, List<List<Integer>> res, List<Integer> cursets) {
        res.add(new ArrayList<>(cursets)); // Add current subset

        for (int i = index; i < arr.length; i++) {
            cursets.add(arr[i]); // Include element
            dfs(arr, i + 1, res, cursets); // Recursive call (use `i + 1`)
            cursets.remove(cursets.size() - 1); // Backtrack
        }

        return res;
    }
    public List<List<Integer>> subsets(int[] nums) {
        return dfs(nums, 0, new ArrayList<>(), new ArrayList<>());
    }
}

File: Data Structures & Algorithms/permutations/submission-0.js
class Solution {
    /**
     * @param {number[]} nums
     * @return {number[][]}
     */
    permute(nums) {
        let res = []
        function backtrack(up,p){
            if(up.length === 0){
                res.push([...p])
                return
            }
            for(let i=0;i<up.length;i++){
                let num = up[i]
                backtrack([...up.slice(0,i),...up.slice(i+1,up.length)],[...p,num])
            }
        }
        backtrack(nums,[])
        return res
    }
}
